README.md
# flutter_date_picker

## Cupertino styled date and time picker component which works on both ios and android.

[![Pub](https://img.shields.io/pub/v/flutter_date_picker.svg)](https://pub.dartlang.org/packages/flutter_date_picker)

# Example

For example on how to use the component [repo](https://github.com/rajeshzmoke/flutter_date_picker/tree/master/example).

## Features

- Seperate Date picker Component.
- Seperate Time picker Component.
- CustomShape for Selected Item.
- CustomGradient for Submit Button.
- CustomItemColor for Selected Item.

<div style="text-align: center">
  <table>
    <tr>
      <td style="text-align: center">
        <img src="https://github.com/rajeshzmoke/flutter_date_picker/blob/master/screenshot/date_time_blue.gif" width="200" />
      </td>
      <td style="text-align: center">
        <img src="https://github.com/rajeshzmoke/flutter_date_picker/blob/master/screenshot/date_time_green.gif" width="200" />
      </td>
      <td style="text-align: center">
        <img src="https://github.com/rajeshzmoke/flutter_date_picker/blob/master/screenshot/date_time_brown.gif" width="200" />
      </td>
    </tr>
  </table>
 </div>

## Getting started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  flutter_date_picker: "^0.1.8"
```

In your library add the following import:

```dart
import 'package:flutter_date_picker/flutter_date_picker.dart';
import 'package:flutter_date_picker/flutter_time_picker.dart';
```

For help getting started with Flutter, view the online [documentation](https://flutter.io/).

# Usage

Let's demo the basic usage with CustomItemColor, CustomShape and CustomGradient

```dart
import 'package:flutter/material.dart';
import 'package:flutter_date_picker/flutter_date_picker.dart';
import 'package:flutter_date_picker/flutter_time_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey<ScaffoldState>();

  VoidCallback _showBottomSheetCallback;
  bool showDatePicker = false;

  @override
  void initState() {
    super.initState();

    _showBottomSheetCallback = _showBottomSheet;
  }

  void _snackBar(String text) {
    _scaffoldKey.currentState.showSnackBar(SnackBar(content: Text(text)));
  }

   _showBottomSheet(String picker) {
    setState(() {
      // disable the button
      _showBottomSheetCallback = null;
    });
    _scaffoldKey.currentState
        .showBottomSheet<Null>(
          (BuildContext context) {
            switch (picker) {
              case 'DatePicker':
                return DatePicker(
                  key: dobKey,
                  setDate: _setDate,
                  customShape: StadiumBorder(
                      side: BorderSide(
                    color: Color(0xFFF991A0),
                  )),
                );
              case 'TimePicker':
                return TimePicker(
                  key: tobKey,
                  setTime: _setTime,
                  customShape: BeveledRectangleBorder(
                    borderRadius: BorderRadius.only(
                      topLeft: Radius.circular(15.0),
                      topRight: Radius.circular(15.0),
                      bottomLeft: Radius.circular(15.0),
                      bottomRight: Radius.circular(15.0),
                    ),
                  ),
                );
            }
          },
        )
        .closed
        .whenComplete(() {
          if (mounted) {
            setState(() {
              // re-enable the button
              _showBottomSheetCallback = _showBottomSheet;
            });
          }
        });
  }

  void _setDate() {
    Navigator.of(context).pop();

    _snackBar(dobKey.currentState.dobStrMonth +
        ' ${dobKey.currentState.dobDate}' +
        ' ${dobKey.currentState.dobYear}');
  }

  void _setTime() {
    Navigator.of(context).pop();
    var hour = tobKey.currentState.hour;

    var min = tobKey.currentState.minute;

    var sec = tobKey.currentState.seconds;

    _snackBar("$hour:$min:$sec ${tobKey.currentState.sunOrMoon}");
  }

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      bottom: false,
      child: Scaffold(
        key: _scaffoldKey,
        appBar: AppBar(
          backgroundColor: Color(0xFFF98495),
          title: Text('Date Picker'),
        ),
        body: Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: <Widget>[
            Container(
              padding: EdgeInsets.all(10.0),
              alignment: Alignment.topCenter,
              child: RaisedButton(
                color: Color(0xFFF98495),
                child: Text('Date Picker'),
                onPressed: () {
                  _showBottomSheet('DatePicker');
                },
              ),
            ),
            Container(
              padding: EdgeInsets.all(10.0),
              alignment: Alignment.topCenter,
              child: RaisedButton(
                color: Color(0xFFF98495),
                child: Text('Time Picker'),
                onPressed: () {
                  _showBottomSheet('TimePicker');
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

## Changelog

Please see the [Changelog](https://github.com/rajeshzmoke/flutter_date_picker/blob/master/CHANGELOG.md) page to know what's recently changed.

## License

```
Copyright 2018 Rajesh Kumar

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
## 0.1.8

- Minor type fix

## 0.1.7

- readme changes

## 0.1.6

- minor code change
- time format

## 0.1.5

- environment changes

## 0.1.4

- environment changes

## 0.1.3

- Added Time Picker Component (hr:min:sec)

## 0.1.2

- Changed the Picker UI to make it more easy to use.
- can add desired background color to selected item.
- can add custom shape for the selectem item.
- can add custom gradient for the submit button.

## 0.1.1

- Initial release.


test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_cupertino_date_picker/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(new MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/date_picker.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

const List<String> monthNames = const <String>[
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

class DatePickerComponent extends StatelessWidget {
  const DatePickerComponent({
    Key key,
    @required this.setMonth,
    @required this.setDate,
    @required this.setYear,
    @required this.themeData,
    @required this.scrollController,
  }) : super(key: key);
  final Function setMonth;
  final Function setDate;
  final Function setYear;
  final ThemeData themeData;
  final FixedExtentScrollController scrollController;

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      children: <Widget>[
        Expanded(
          flex: 1,
          child: Container(
              padding: EdgeInsets.all(8.0),
              height: 150.0,
              // width: 150.0,
              decoration: BoxDecoration(
                  border:
                      Border(top: BorderSide(color: themeData.disabledColor)),
                  color: Colors.white),
              child: CupertinoPicker(
                backgroundColor: Colors.white,
                scrollController: scrollController,
                itemExtent: 28.0,
                onSelectedItemChanged: (int index) {
                  setMonth(index);
                },
                children: List<Widget>.generate(monthNames.length, (int index) {
                  return Text(
                    monthNames[index],
                    style: TextStyle(color: Color(0xFF000046), fontSize: 20.0),
                    textAlign: TextAlign.start,
                  );
                }),
              )),
        ),
        Expanded(
          flex: 1,
          child: Container(
              padding: EdgeInsets.all(8.0),
              height: 150.0,
              // width: 150.0,
              decoration: BoxDecoration(
                  border:
                      Border(top: BorderSide(color: themeData.disabledColor)),
                  color: Colors.white),
              child: CupertinoPicker(
                backgroundColor: Colors.white,
                scrollController: scrollController,
                itemExtent: 28.0,
                onSelectedItemChanged: (int index) {
                  setDate(index);
                },
                children: List<Widget>.generate(31, (int index) {
                  return Text(
                    "${index + 1}",
                    style: TextStyle(color: Color(0xFF000046), fontSize: 20.0),
                    textAlign: TextAlign.start,
                  );
                }),
              )),
        ),
        Expanded(
          flex: 1,
          child: Container(
              padding: EdgeInsets.all(8.0),
              height: 150.0,
              // width: 150.0,
              decoration: BoxDecoration(
                  border:
                      Border(top: BorderSide(color: themeData.disabledColor)),
                  color: Colors.white),
              child: CupertinoPicker(
                backgroundColor: Colors.white,
                scrollController: scrollController,
                itemExtent: 28.0,
                onSelectedItemChanged: (int index) {
                  setYear(index);
                },
                children: List<Widget>.generate(69, (int index) {
                  return Text(
                    '${1950 + index}',
                    style: TextStyle(color: Color(0xFF000046), fontSize: 20.0),
                    textAlign: TextAlign.start,
                  );
                }),
              )),
        ),
      ],
    );
  }
}


lib/flutter_date_picker.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_date_picker/dob_cupertino_picker.dart';

const List<String> monthNames = const <String>[
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];

const LinearGradient zGradient =
    const LinearGradient(begin: Alignment(-0.5, 1.0), colors: [
  Color(0xFFF98495),
  Color(0xFF595DC2),
]);

/// Color of the item background
const Color zItemColor = const Color(0xFFF98495);

/// Shape of the item
const ShapeBorder zShape = const BeveledRectangleBorder(
  borderRadius: BorderRadius.only(
    topLeft: Radius.circular(15.0),
    bottomRight: Radius.circular(15.0),
  ),
);

///Global key to access the getters
final dobKey = GlobalKey<DatePickerState>();

class DatePicker extends StatefulWidget {
  DatePicker({
    Key key,
    @required this.setDate,
    this.customItemColor = zItemColor,
    this.customShape = zShape,
    this.customGradient = zGradient,
  }) : super(key: key);

  /// Function to get the data
  final Function setDate;

  /// Custom LinerGradient for the submit button
  final LinearGradient customGradient;

  /// BackgroundColor for the selected item (Month/date/year)
  final Color customItemColor;

  /// Custom shape for the selected item
  final ShapeBorder customShape;

  @override
  DatePickerState createState() => DatePickerState();
}

class DatePickerState extends State<DatePicker> {
  DateTime _date = DateTime.now();

  int _selectedMonth;
  int _selectedDate;
  int _selectedYear;
  String _dobMonth;

  int get dobMonth => _selectedMonth;
  String get dobDate =>
      _selectedDate + 1 < 10 ? '0${_selectedDate + 1}' : '${_selectedDate + 1}';
  int get dobYear => (_date.year - 118) + _selectedYear;
  String get dobStrMonth => _dobMonth;

  @override
  void initState() {
    super.initState();
    _selectedMonth = _date.month - 1;
    _selectedDate = _date.day - 1;
    _selectedYear = 74;

    _dobMonth = monthNames[_date.month - 1];
  }

  @override
  Widget build(BuildContext context) {
    final ThemeData themeData = Theme.of(context);
    return Column(
      mainAxisAlignment: MainAxisAlignment.end,
      children: <Widget>[
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Expanded(
              flex: 1,
              child: Container(
                padding: EdgeInsets.only(left: 18.0),
                height: 150.0,
                decoration: BoxDecoration(
                    border: Border(
                        top: BorderSide(
                      color: themeData.disabledColor,
                    )),
                    color: Colors.white),
                child: Container(
                  margin: EdgeInsets.symmetric(horizontal: 8.0),
                  child: CustomCupertinoPicker(
                    customShape: widget.customShape,
                    customColor: widget.customItemColor,
                    backgroundColor: Colors.white,
                    scrollController: FixedExtentScrollController(
                      initialItem: _selectedMonth,
                    ),
                    itemExtent: 35.0,
                    diameterRatio: 1.2,
                    onSelectedItemChanged: (int index) {
                      setState(() {
                        _selectedMonth = index;
                        _dobMonth = monthNames[index];
                      });
                    },
                    children:
                        List<Widget>.generate(monthNames.length, (int index) {
                      return Padding(
                        padding: const EdgeInsets.symmetric(vertical: 6.0),
                        child: Text(
                          monthNames[index],
                          style: TextStyle(
                            color: _selectedMonth == index
                                ? Colors.white
                                : widget.customItemColor.withOpacity(0.6),
                            fontSize: 18.0,
                            fontWeight: FontWeight.w600,
                          ),
                          textAlign: TextAlign.center,
                        ),
                      );
                    }),
                  ),
                ),
              ),
            ),
            Expanded(
              flex: 1,
              child: Container(
                  padding: EdgeInsets.only(left: 18.0),
                  height: 150.0,
                  decoration: BoxDecoration(
                      border: Border(
                          top: BorderSide(
                        color: themeData.disabledColor,
                      )),
                      color: Colors.white),
                  child: Container(
                    margin: EdgeInsets.symmetric(horizontal: 8.0),
                    child: CustomCupertinoPicker(
                      customShape: widget.customShape,
                      customColor: widget.customItemColor,
                      backgroundColor: Colors.white,
                      scrollController: FixedExtentScrollController(
                        initialItem: _selectedDate,
                      ),
                      itemExtent: 35.0,
                      diameterRatio: 1.2,
                      onSelectedItemChanged: (int index) {
                        setState(() {
                          _selectedDate = index;
                        });
                      },
                      children: List<Widget>.generate(31, (int index) {
                        return Padding(
                          padding: const EdgeInsets.symmetric(vertical: 6.0),
                          child: Text(
                            "${index + 1}",
                            style: TextStyle(
                              color: _selectedDate == index
                                  ? Colors.white
                                  : widget.customItemColor.withOpacity(0.6),
                              fontSize: 18.0,
                              fontWeight: FontWeight.w600,
                            ),
                            textAlign: TextAlign.start,
                          ),
                        );
                      }),
                    ),
                  )),
            ),
            Expanded(
              flex: 1,
              child: Container(
                  padding: EdgeInsets.only(left: 18.0),
                  height: 150.0,
                  decoration: BoxDecoration(
                    border:
                        Border(top: BorderSide(color: themeData.disabledColor)),
                    color: Colors.white,
                  ),
                  child: Container(
                    margin: EdgeInsets.symmetric(horizontal: 10.0),
                    child: CustomCupertinoPicker(
                      customShape: widget.customShape,
                      customColor: widget.customItemColor,
                      backgroundColor: Colors.white,
                      scrollController: FixedExtentScrollController(
                        initialItem: _selectedYear,
                      ),
                      itemExtent: 35.0,
                      diameterRatio: 1.2,
                      onSelectedItemChanged: (int index) {
                        setState(() {
                          _selectedYear = index;
                        });
                      },
                      children: List<Widget>.generate(118, (int index) {
                        return Padding(
                          padding: const EdgeInsets.symmetric(vertical: 6.0),
                          child: Text(
                            '${(_date.year - 118) + index}',
                            style: TextStyle(
                              color: _selectedYear == index
                                  ? Colors.white
                                  : widget.customItemColor.withOpacity(0.6),
                              fontSize: 18.0,
                              fontFamily: 'Nunito',
                              fontWeight: FontWeight.w600,
                            ),
                            textAlign: TextAlign.start,
                          ),
                        );
                      }),
                    ),
                  )),
            ),

            /// Submit button
            Expanded(
              flex: 1,
              child: Container(
                height: 150.0,
                decoration: BoxDecoration(
                    border:
                        Border(top: BorderSide(color: themeData.disabledColor)),
                    color: Colors.white),
                child: Center(
                  child: ClipOval(
                    child: Container(
                      width: 50.0,
                      height: 50.0,
                      decoration: BoxDecoration(
                        border: Border(
                            top: BorderSide(
                          color: themeData.disabledColor,
                        )),
                        gradient: widget.customGradient ?? zGradient,
                      ),
                      child: IconButton(
                        color: Colors.white,
                        icon: Icon(
                          Icons.send,
                          size: 20.0,
                        ),
                        onPressed: () {
                          widget.setDate();
                        },
                      ),
                    ),
                  ),
                ),
              ),
            )
          ],
        ),
      ],
    );
  }
}


lib/dob_cupertino_picker.dart
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/foundation.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart';

/// Color of the 'magnifier' lens border.

const Color _kDefaultBackground = const Color(0xFFD2D4DB);

/// Eyeballed value comparing with a native picker.
const double _kDefaultDiameterRatio = 1.1;

/// Opacity fraction value that hides the wheel above and below the 'magnifier'
/// lens with the same color as the background.
const double _kForegroundScreenOpacityFraction = 0.7;

/// An iOS-styled picker.
///
/// Displays the provided [children] widgets on a wheel for selection and
/// calls back when the currently selected item changes.
///
/// Can be used with [showModalBottomSheet] to display the picker modally at the
/// bottom of the screen.
///
/// See also:
///
///  * [ListWheelScrollView], the generic widget backing this picker without
///    the iOS design specific chrome.
///  * <https://developer.apple.com/ios/human-interface-guidelines/controls/pickers/>
class CustomCupertinoPicker extends StatefulWidget {
  /// Creates a control used for selecting values.
  ///
  /// The [diameterRatio] and [itemExtent] arguments must not be null. The
  /// [itemExtent] must be greater than zero.
  ///
  /// The [backgroundColor] defaults to light gray. It can be set to null to
  /// disable the background painting entirely; this is mildly more efficient
  /// than using [Colors.transparent].
  const CustomCupertinoPicker({
    Key key,
    this.diameterRatio = _kDefaultDiameterRatio,
    this.backgroundColor = _kDefaultBackground,
    this.scrollController,
    this.customColor,
    this.customShape,
    @required this.itemExtent,
    @required this.onSelectedItemChanged,
    @required this.children,
  })  : assert(diameterRatio != null),
        assert(diameterRatio > 0.0,
            RenderListWheelViewport.diameterRatioZeroMessage),
        assert(itemExtent != null),
        assert(itemExtent > 0),
        super(key: key);

  /// BackgroundColor for the selected item (Month/date/year)

  final Color customColor;

  /// Custom shape for the selected item
  final ShapeBorder customShape;

  /// Relative ratio between this picker's height and the simulated cylinder's diameter.
  ///
  /// Smaller values creates more pronounced curvatures in the scrollable wheel.
  ///
  /// For more details, see [ListWheelScrollView.diameterRatio].
  ///
  /// Must not be null and defaults to `1.1` to visually mimic iOS.
  final double diameterRatio;

  /// Background color behind the children.
  ///
  /// Defaults to a gray color in the iOS color palette.
  ///
  /// This can be set to null to disable the background painting entirely; this
  /// is mildly more efficient than using [Colors.transparent].
  final Color backgroundColor;

  /// A [FixedExtentScrollController] to read and control the current item.
  ///
  /// If null, an implicit one will be created internally.
  final FixedExtentScrollController scrollController;

  /// The uniform height of all children.
  ///
  /// All children will be given the [BoxConstraints] to match this exact
  /// height. Must not be null and must be positive.
  final double itemExtent;

  /// An option callback when the currently centered item changes.
  ///
  /// Value changes when the item closest to the center changes.
  ///
  /// This can be called during scrolls and during ballistic flings. To get the
  /// value only when the scrolling settles, use a [NotificationListener],
  /// listen for [ScrollEndNotification] and read its [FixedExtentMetrics].
  final ValueChanged<int> onSelectedItemChanged;

  /// [Widget]s in the picker's scroll wheel.
  final List<Widget> children;

  @override
  State<StatefulWidget> createState() => _CupertinoPickerState();
}

class _CupertinoPickerState extends State<CustomCupertinoPicker> {
  int _lastHapticIndex;

  void _handleSelectedItemChanged(int index) {
    // Only the haptic engine hardware on iOS devices would produce the
    // intended effects.
    if (defaultTargetPlatform == TargetPlatform.iOS &&
        index != _lastHapticIndex) {
      _lastHapticIndex = index;
      HapticFeedback.selectionClick();
    }

    if (widget.onSelectedItemChanged != null) {
      widget.onSelectedItemChanged(index);
    }
  }

  /// Makes the fade to white edge gradients.
  Widget _buildGradientScreen() {
    return Positioned.fill(
      child: IgnorePointer(
        child: Container(
          decoration: const BoxDecoration(
            gradient: const LinearGradient(
              colors: const <Color>[
                const Color(0xFFFFFFFF),
                const Color(0xF2FFFFFF),
                const Color(0xDDFFFFFF),
                const Color(0x00FFFFFF),
                const Color(0x00FFFFFF),
                const Color(0xDDFFFFFF),
                const Color(0xF2FFFFFF),
                const Color(0xFFFFFFFF),
              ],
              stops: const <double>[
                0.0,
                0.05,
                0.09,
                0.22,
                0.78,
                0.91,
                0.95,
                1.0,
              ],
              begin: Alignment.topCenter,
              end: Alignment.bottomCenter,
            ),
          ),
        ),
      ),
    );
  }

  /// Makes the magnifier lens look so that the colors are normal through
  /// the lens and partially grayed out around it.
  Widget _buildMagnifierScreen() {
    final Color foreground = widget.backgroundColor?.withAlpha(
        (widget.backgroundColor.alpha * _kForegroundScreenOpacityFraction)
            .toInt());

    return IgnorePointer(
      child: Column(
        children: <Widget>[
          Expanded(
            child: Container(
              color: foreground,
            ),
          ),
          Container(
            decoration: ShapeDecoration(
              shape: widget.customShape,
              //  StadiumBorder(
              //   side: BorderSide(
              //     color: widget.customColor,
              //   ),
              // ),
              color: widget.customColor,
            ),
            constraints: BoxConstraints.expand(height: widget.itemExtent),
          ),
          Expanded(
            child: Container(
              color: foreground,
            ),
          ),
        ],
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    Widget result = Stack(
      children: <Widget>[
        _buildMagnifierScreen(),
        Positioned.fill(
          child: ListWheelScrollView(
            controller: widget.scrollController,
            physics: const FixedExtentScrollPhysics(),
            diameterRatio: widget.diameterRatio,
            itemExtent: widget.itemExtent,
            onSelectedItemChanged: _handleSelectedItemChanged,
            children: widget.children,
          ),
        ),
        _buildGradientScreen(),
      ],
    );
    if (widget.backgroundColor != null) {
      result = DecoratedBox(
        decoration: BoxDecoration(
          color: widget.backgroundColor,
        ),
        child: result,
      );
    }
    return result;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_date_picker/flutter_date_picker.dart';
import 'package:flutter_date_picker/flutter_time_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey<ScaffoldState>();

  Function _showBottomSheetCallback;
  bool showDatePicker = false;

  @override
  void initState() {
    super.initState();

    _showBottomSheetCallback = _showBottomSheet;
  }

  void _snackBar(String text) {
    _scaffoldKey.currentState.showSnackBar(SnackBar(content: Text(text)));
  }

  _showBottomSheet(String picker) {
    setState(() {
      // disable the button
      _showBottomSheetCallback = null;
    });
    _scaffoldKey.currentState
        .showBottomSheet<Null>(
          (BuildContext context) {
            switch (picker) {
              case 'DatePicker':
                return DatePicker(
                  key: dobKey,
                  setDate: _setDate,
                  customShape: BeveledRectangleBorder(
                    borderRadius: BorderRadius.only(
                      topLeft: Radius.circular(15.0),
                      topRight: Radius.circular(15.0),
                      bottomLeft: Radius.circular(15.0),
                      bottomRight: Radius.circular(15.0),
                    ),
                  ),
                  customItemColor: Color(0xFFba5905),
                  customGradient:
                      LinearGradient(begin: Alignment(-0.5, 1.0), colors: [
                    Color(0xFFba5905),
                    Color(0xFFefcaaa),
                    Color(0xFFba5905),
                  ]),
                );
              case 'TimePicker':
                return TimePicker(
                  key: tobKey,
                  setTime: _setTime,
                  customShape: StadiumBorder(
                      side: BorderSide(
                    color: Color(0xFFba5905),
                  )),
                  customItemColor: Color(0xFFba5905),
                  customGradient:
                      LinearGradient(begin: Alignment(-0.5, 1.0), colors: [
                    Color(0xFFba5905),
                    Color(0xFFefcaaa),
                    Color(0xFFba5905),
                  ]),
                );
            }
          },
        )
        .closed
        .whenComplete(() {
          if (mounted) {
            setState(() {
              // re-enable the button
              _showBottomSheetCallback = _showBottomSheet;
            });
          }
        });
  }

  void _setDate() {
    Navigator.of(context).pop();

    _snackBar(dobKey.currentState.dobStrMonth +
        ' ${dobKey.currentState.dobDate}' +
        ' ${dobKey.currentState.dobYear}');
  }

  void _setTime() {
    Navigator.of(context).pop();
    var hour = tobKey.currentState.hour;

    var min = tobKey.currentState.minute;

    var sec = tobKey.currentState.seconds;

    _snackBar("$hour:$min:$sec ${tobKey.currentState.sunOrMoon}");
  }

  @override
  Widget build(BuildContext context) {
    return SafeArea(
      bottom: false,
      child: Scaffold(
        key: _scaffoldKey,
        appBar: AppBar(
          backgroundColor: Color(0xFFba5905),
          title: Text('Date and Time Picker'),
        ),
        body: Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: <Widget>[
            Container(
              padding: EdgeInsets.all(10.0),
              alignment: Alignment.topCenter,
              child: RaisedButton(
                textColor: Colors.white,
                color: Color(0xFFba5905),
                child: Text('Date Picker'),
                onPressed: () {
                  _showBottomSheet('DatePicker');
                },
              ),
            ),
            Container(
              padding: EdgeInsets.all(10.0),
              alignment: Alignment.topCenter,
              child: RaisedButton(
                textColor: Colors.white,
                color: Color(0xFFba5905),
                child: Text('Time Picker'),
                onPressed: () {
                  _showBottomSheet('TimePicker');
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}
